sns_ik_lib/src/sns_ik_math_utils.cpp

```diff
@@ -228,4 +229,50 @@ bool isIdentity(const Eigen::MatrixXd &A) {
   return isIdentity;
 }
 
+/*************************************************************************************************/
+
+void pseudoInverse(const Eigen::MatrixXd& A, double eps, Eigen::MatrixXd* invA,
+                   int* rank, bool* damped)
+{
+  // Compute the singular value decomposition:
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd_A(A.transpose(), Eigen::ComputeThinU | Eigen::ComputeThinV);
+  Eigen::VectorXd sigma = svd_A.singularValues();
+
+  // Compute the rank by checking for non-zero singular values
+  int rankA = 0;
+  int nSigma = sigma.size();
+  for (int i = 0; i < nSigma; i++) {
+    if (sigma(i) > eps) rankA++;  // increment the rank counter
```

Comment:
Do we care if `eps` is negative and therefore `sigma(i) == 0`?

Reply:
Good point - we should do input validation on `eps` to ensure that it is positive.